[inductor] Add scatter backend for large max_pool2d kernels

stmcgovern opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* (to be filled)

Fixes #66042: Eliminates eager fallback for large kernels (window_size > 25)
by implementing O(output_size) scatter-based backward pass.

Changes:
• Add _max_pool2d_scatter_backward() with scatter_add-based gradient accumulation
• Add _validate_max_pool2d_params() for robust parameter validation
• Enable dilation support for scatter backend (was previously blocked)
• Add intelligent backend selection: scatter (large) vs optimized (small)
• Add comprehensive error handling with graceful fallback
• Add 6 test methods in test_torchinductor.py for regression prevention

Performance claim (to be verified): Eliminates O(kernel_size²) complexity and eager fallback.
Significant speedup for large kernels, zero regression for small kernels.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @chenyang78

isuruf commented on torch/_inductor/lowering.py:
```diff
@@ -4742,10 +4742,215 @@ def max_pool3d_with_indices(
+            0, grad_output.get_device(), grad_output.get_dtype(), [x.get_numel()]
+        )
+
+        # Handle invalid indices (-1) by replacing with valid index and zeroing values
+        valid_mask = ops.ne(scatter_indices_flat, ops.constant(-1, torch.int64))
```
`ops.*` is supposed to be for used in functions passed to `inner_fn` in `ir.Pointwise.create`. I'm not sure how it works here.

isuruf commented on torch/_inductor/lowering.py:
```diff
@@ -4742,10 +4742,215 @@ def max_pool3d_with_indices(
+        # Reshape back to input dimensions
+        return view(result_flat, x.get_size())
+
+    except (TypeError, AttributeError):
```
Let's remove these try except and figure out under what conditions these exceptions are thrown and check for those conditions at the beginning.

isuruf reviewed (commented):
It's a good improvment. I'm wondering if we should write a traditional `O(output_size * kernel_size**2)` lowering for large kernel sizes. Then we can fallback to that triton kernel instead of the eager fallback. (This can be a separate PR).

stmcgovern committed:
[inductor] Add scatter backend for large max_pool2d backward kernels

Fixes #66042: Eliminates eager fallback for large max_pool2d kernels (window_size > 25)
by implementing an O(output_size) scatter-based backward pass instead of the traditional
O(output_size × kernel_size²) approach.

- Add `_max_pool2d_scatter_backward()`: New scatter-based backward implementation
  - Transforms max_pool2d indices (linear H×W positions) to full 4D tensor indices
  - Uses scatter_add for atomic gradient accumulation
  - Handles invalid indices with bounds checking and masking
  - Complexity: O(output_size) vs O(output_size × kernel_size²) for traditional approach

- Update `max_pool2d_with_indices_backward()`: Intelligent backend dispatch
  - Large kernels (window_size > 25, !ceil_mode): Use scatter backend
  - Small kernels: Use existing optimized Triton path
  - Dilation support: Scatter backend supports all dilations; small-kernel path falls back for dilation != 1
  - ceil_mode: Not supported by scatter backend, falls back to eager

- Remove max_pool2d_with_indices_backward from decompositions to enable custom lowering

- Add 3 regression tests:
  - `test_max_pool2d_scatter_backward_issue_66042`: Basic functionality for large kernels
  - `test_max_pool2d_scatter_correctness_large_kernels`: Gradient correctness across kernel sizes
  - `test_max_pool2d_scatter_overlapping_windows`: Overlapping receptive fields edge case

The scatter backend eliminates the O(kernel_size²) iteration over the pooling window by:
1. Converting the pooled indices (which already identify the max position) directly to scatter targets
2. Using atomic scatter_add for gradient accumulation
3. Avoiding the expensive backward window iteration

For large kernels (e.g., 13×13 = 169 elements), this provides significant speedup while maintaining correctness.

- Graceful fallback for edge cases (ceil_mode, etc.)

stmcgovern commented:
I'm working on understanding the numerical differences in these two tests.

```
_________________________________________________ GPUTests.test_max_pool2d_with_indices_backward5_cuda __________________________________________________
Traceback (most recent call last):
  File "/miniconda/lib/python3.12/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/miniconda/lib/python3.12/unittest/case.py", line 634, in run
    self._callTestMethod(testMethod)
  File "/miniconda/lib/python3.12/unittest/case.py", line 589, in _callTestMethod
    if method() is not None:
       ^^^^^^^^
  File "/workspace/pytorch/torch/testing/_internal/common_utils.py", line 3300, in wrapper
    method(*args, **kwargs)
  File "/workspace/pytorch/test/inductor/test_torchinductor.py", line 14473, in new_test
    return value(self)
           ^^^^^^^^^^^
  File "/workspace/pytorch/torch/testing/_internal/common_device_type.py", line 1419, in efail_fn
    return fn(slf, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/pytorch/test/inductor/test_torchinductor.py", line 9709, in test_max_pool2d_with_indices_backward5
    self.common(
  File "/miniconda/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/workspace/pytorch/test/inductor/test_torchinductor.py", line 720, in check_model_gpu
    check_model(
  File "/workspace/pytorch/test/inductor/test_torchinductor.py", line 561, in check_model
    assert_equal_fn(
  File "/workspace/pytorch/torch/_dynamo/test_case.py", line 112, in assertEqual
    return super().assertEqual(x, y, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/pytorch/torch/testing/_internal/common_utils.py", line 4255, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Tensor-likes are not close!

Mismatched elements: 177 / 51200 (0.3%)
Greatest absolute difference: 0.046875 at index (0, 59, 5, 14) (up to 1e-05 allowed)
Greatest relative difference: 0.46630859375 at index (1, 57, 5, 16) (up to 0.001 allowed)

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor.py GPUTests.test_max_pool2d_with_indices_backward5_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
----------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------
stats [('calls_captured', 2), ('unique_graphs', 2)]
aot_autograd [('total', 2), ('autograd_cache_miss', 2), ('autograd_cache_saved', 2), ('ok', 2)]
inductor [('triton_bundler_save_kernel', 32), ('async_compile_cache_miss', 8), ('async_compile_cache_hit', 4), ('fxgraph_cache_miss', 2), ('triton_bundler_save_static_autotuner', 2)]
graph_break []
================================================================ short test summary info ================================================================
FAILED [1.6098s] test/inductor/test_torchinductor.py::GPUTests::test_max_pool2d_with_indices_backward2_cuda - AssertionError: Tensor-likes are not close!
FAILED [0.3896s] test/inductor/test_torchinductor.py::GPUTests::test_max_pool2d_with_indices_backward5_cuda - AssertionError: Tensor-likes are not close!
=============================================== 2 failed, 58 passed, 1870 deselected in 310.26s (0:05:10) ===============================================
(base) [root@c05d4db1e639 pytorch]# 
```

stmcgovern commented:
Summary of test failures

TODO lint, label and xfail removal since `test_max_pool2d_with_indices_backward5_dynamic_shapes_cpu` passes now. 

local (not sure when CI runs them)
- There is a history for the test_max_pool2d_with_indices_backward2 with ceil_mode. Not sure this PR should try to address that. Consider turning back to eager (for now). Can be addressed in follow on [PR](https://github.com/pytorch/pytorch/pull/166662) with traditional backend. 

- It seems that test_max_pool2d_with_indices_backward5_cuda  (13x13 kernel) is likely due to gpu atomics nondeterminism. Consider relaxing tolerances to something acceptable.

eellison reviewed (commented):
Thank you for the pr! Handling dynamic shapes in poolingCan I ask what the advantage of this is instead of doing a gather + reduction ?  

Deferring to @isuruf for review unless you want another look

eellison commented:
Yea, I think for the backward, scatter makes more sense. I was conflating the two. If it's easy enough, could do both and benchmark. Although for fwd I do think we want gather + reduction.

isuruf commented:
>  Although for fwd I do think we want gather + reduction.

Yes, we do have a gather + reduction for the forward implementation already.

eellison commented:
> Yes, we do have a gather + reduction for the forward implementation already.

I had a specific ask for avg_pool2d with dynamic kernel size. This is pretty easily implementable as a decomp.. Although potentially we could do lowering as well. Do you mind if I submit the pr ?

isuruf commented:
Sure. Here's two stale PRs where you could have a look
- Lowering as a reduction: https://github.com/pytorch/pytorch/pull/116756
- Decomposition: https://github.com/pytorch/pytorch/pull/116418
